Pull from upstream

dgonyeo opened the pull request:
There were many, many conflicts here.

Notable new changes include (but are not limited to):

- s/ncurses/libuncursed/ A new rendering library. Supports different
  tilesets (ascii, unicode, and some graphical tiles)
- A new in-game window displays possible actions a player can take
- Rewritten save system that can handle watching and replaying games
- "lots and lots of bugfixes"
- cmake is dead, long live aimake (upstream dropped support for cmake,
  so we will too)
- Multi-turn actions (like reading a spellbook or shift-moving) will now
  step through all the turns rapidly, instead of displaying them as a
  single turn
- Various UI tweaks, like the window borders turn red now when you're
  low on health

Some of the things that caused the biggest conflicts:

- __Ring of Power__: It needed to be largely reimplemented. Tracking what your state with the ring is (none, carrying, or being worn) is now done through extrinsics the ring provides. Previously it was done by adding fields to the user's struct.
- __Trident of Fish Command__: The way The Trident of Fish Command detected fish no longer works. Now the trident provides an extrinsic, which is checked for in the same place ESP, telepathy, etc. are checked.
- __Drunken Boxing__: The file (libnethack/src/uhitm.c) this was implemented in appears to have been completely rewritten. I grabbed the diffs from our changes, and attempted to work the logic back in to the new file. I think it works, but others testing this would be appreciated
- __aimake__: With aimake being the One Build System to rule them all, it needed tweaking to support Potter's rumors file. Also we're currently lacking tiles for Potter, Lilith, Lolth, Mammon, Chulhu, and Sammael. This only produces a warning, and I intend to create these tiles for a later PR.

Building now works as follows:
```bash
mkdir build
cd build
../aimake -i /tmp/bingehack4 -v
```

It will install itself into the `/tmp/bingehack4` directory.

dgonyeo commented:
Please assume these changes are horribly broken and require many fixes. A lot of the merge conflicts I resolved I wasn't completely sure about. I've tested what I can think of, but chances are I fucked something up somewhere.

dgonyeo commented:
Upstream is currently in the middle of 4.3-beta2. I intend to pull from them regularly after this.

eatnumber1 commented:
This is virtually impossible to review. Can you either:

1. Pull out your changes and make this purely a merge commit that can be approved without review, then in a separate commit introduce your changes, or
2. Add Github comments to the specific files or sections of code you want me to review

dgonyeo commented:
I'm going to go with option #2, since parts of the game will be broken with the merge conflict resolutions and not my changes.

dgonyeo commented on libnethack/dat/gehennom.des:
```diff
@@ -205,11 +206,11 @@ RANDOM_PLACES:(04,02),(46,02),(04,15),(46,15)
-BRANCH:levregion(01,00,11,20),(0,0,50,17)
-TELEPORT_REGION:levregion(01,00,11,20),(0,0,50,17),up
-TELEPORT_REGION:levregion(69,00,79,20),(0,0,50,17),down
+STAIR:levregion(00,00,10,20),(0,0,50,17),down
```
I had to tweak these numbers (it was complaining about being out of range). Please double check that the levels are sane with these changes from here to the rest of the file.

dgonyeo commented on libnethack/dat/medusa.des:
```diff
@@ -48,7 +49,7 @@ DOOR:locked,(38,08)
 DOOR:closed,(30,12)
 # Branch, not allowed inside Medusa's building.
-BRANCH:levregion(01,00,79,20),(30,06,46,13)
+BRANCH:levregion(00,00,78,20),(30,06,46,13)
```
Same as before, I had to decrement this number. Please confirm the level is still sane.

dgonyeo commented on libnethack/include/artifact.h:
```diff
@@ -35,7 +40,9 @@
-# define SPFX_FISH   0x10000000UL
+# define SPFX_FISH   0x10000000UL /* Fish Detection */
+# define SPFX_RPOWER 0x20000000UL /* Has the Ring of Power */
+# define SPFX_WRPOWER 0x40000000UL /* Wearing the Ring of Power */
```
These were added as special effects the Ring of Power can convey.

dgonyeo commented on libnethack/include/artilist.h:
```diff
@@ -225,7 +263,8 @@ static const struct artifact const_artilist[] = {
     A("The Ring of Power", RIN_SLOW_DIGESTION,
-      (SPFX_RESTR | SPFX_INTEL | SPFX_INVIS), 0, 0, //It speaks, it is just handled elsewhere (allmain)
+      (SPFX_RESTR | SPFX_INTEL | SPFX_INVIS | SPFX_WRPOWER),
+      SPFX_RPOWER, 0, //It speaks, it is just handled elsewhere (allmain)
```
The ring of power now conveys `SPFX_RPOWER` while being carried, and `SPFX_WRPOWER` while being worn.

dgonyeo commented on libnethack/include/decl.h:
```diff
@@ -319,99 +255,144 @@ struct histevent {
+# define lilith_level           (dungeon_topology.d_lilith_level)
+# define sammael_level          (dungeon_topology.d_sammael_level)
+# define cthulhu_level          (dungeon_topology.d_cthulhu_level)
+# define mammon_level           (dungeon_topology.d_mammon_level)
```
I had to add these `# define`s to make the levels we had added work.

dgonyeo commented on libnethack/include/display.h:
```diff
@@ -7,126 +8,168 @@
+# define MSENSE_SCENT         0x00000200u
+
+/* Detect Fish. For the Trident of Fish */
+# define MSENSE_DETECTFISH    0x00000400u
```
`MSENSE_DETECTFISH` was added for the Trident of Fish Command. It was also added to `MSENSE_ANYVISION` a couple lines down.

dgonyeo commented on libnethack/include/extern.h:
```diff
@@ -533,17 +595,20 @@ extern void display_file(const char *, boolean);
 extern int delete_bonesfile(char *bonesid);
 extern void paniclog(const char *, const char *);
-extern boolean lock_fd(int fd, int retry);
-extern void unlock_fd(int fd);
```
I removed `lock_fd` and `unlock_fd` because they were conflicting, and nothing appeared to be using them anymore.

dgonyeo commented:
It's refusing to show many of the files on the site due to there being too many files that have been edited, so here's the rest of the locations that need looking at:

- `libnethack/include/prop.h`:86, I added properties for detect fish, carrying the ring of power, and wearing the ring of power
- `libnethack/include/youprop.h` All Detect_fish properties were removed.
- `libnethack/src/allmain.c`:950 The checks for if you have / are wearing the Ring of Power were changed to look for the respective properties. They used to check flags on the `u` struct.
- `libnethack/src/artifact.c`:477 The `item_provides_extrinsic` function was updated to check for the detect fish, and ring of power extrinsics.
- `libnethack/src/do_wear.c`:494 The functions for putting on and removing equipment changed, so the logic for these behaviors for the Ring of Power was moved to here
- `libnethack/src/events.c`:36-37 I removed the line `char *str = strerror(errno);` due to `str` not being used
- `libnethack/src/pager.c`:133 Things can be seen via Detect Fish
- `libnethack/src/prop.c`:375 If we have line of sight to a swimmer, and we can detect fish, we can see them.
- `libnethack/src/uhitm.c` All drunken boxing logic here should be checked. Sections of it start at lines 29, 206, and 287.
- `aimake.rules`:307 Potter rumor files were added.

dgonyeo commented:
Also the license files were modified, so those should be glanced at too.

eatnumber1 commented:
Finally got started on this review. I'm only getting started though, but here's what I've got so far:
 * `libnethack/dat/oracles.txt`: Trailing whitespace
 * `libnethack/include/artilist.h`: Indent continuation lines further
 * `libnethack/include/prop.h`: `LAST_PROP` should be `WORN_RING_OF_POWER`
 * `libnethack/include/you.h`: Fix whitespace
 * `libnethack/include/you.h`: Don't litter the code with `#defines`. Instead, put that define in a header, and document what it's for.

dgonyeo commented:
Adding a commit to address some of these things. 

In `libnethack/include/artilist.h` the continuation lines indentation for the artifacts we've added is following what the indentation is from upstream.

In `libnethack/include/you.h` all `#define`s are from upstream.

What part of the whitespace in `libnethack/include/you.h` should be fixed?

dgonyeo commented:
ping

clockfort commented:
/me will start reading the changes here

Not to derail the thread on purpose, but, do we know what upstream is doing with the recent official nethack release? (hereafter referred to as "upstream upstream")

dgonyeo commented:
I didn't look too closely at the changelog, but my understanding was that it was mostly establishing that they will be carrying the banner forward for nethack. I don't know what the implications on the nethack4 team are. 

> On Dec 29, 2015, at 04:50, Chris Lockfort <notifications@github.com> wrote:
> 
> /me will start reading the changes here
> 
> Not to derail the thread on purpose, but, do we know what upstream is doing with the recent official nethack release? (hereafter referred to as "upstream upstream")
> 
> ―
> Reply to this email directly or view it on GitHub.
>

eatnumber1 commented:
Had a look at *just* the files you listed. Those look good to me.

worr commented:
Ugh, github swallowed my email.

Re: nethack 3.6.0:
It's more than just lip service; they merged some of the patches that
have been part of nethack 4.

I guess the question becomes whether or not we want something closer to
vanilla nethack, or if we want to keep the changes from nethack 4.0 if
they're still continuing development.

Unfortunately, upstream upstream's build system still sucks worse than
upstream's.

dgonyeo commented:
fwiw I do think the UI improvements made in upstream over upstream upstream has significant value, especially when indoctrinating new members with bingehack. If upstream upstream merges those patches in however, I don't know where that would leave us.
